### DISCUSSION TITLE
Welcome to emerald Discussions!

### DISCUSSION THREAD

**Discussion by rdmsr** [Abb1x/example-kernel#8] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ghost** [comment]
take my userspace stuff. (C++)

Abb1x <notifications@github.com> schrieb am Di., 8. Dez. 2020, 22:56:

**Comment by wuare** [comment]
i'm new to os, how to run:sweat_smile:？

**Reply by rdmsr** [reply]
You need to compile the OS using a cross compiler and `make run ` to get the latest version or if you don't want to compile it , prebuilt images are available in the "releases" tab, You can run these images in Virtual Box using the "choose a virtual disk" option. I might work on a build guide

**Comment by ghost** [comment]
what OS are you using?

**Reply by dko1905** [reply]
He is using Linux. But it should be possible on windows.

**Reply by rdmsr** [reply]
> He is using Linux. But it should be possible on windows.

yes, it should work in WSL

**Comment by NETLIFYHOSTINGAHMD** [comment]
how can i build and run from src because release is not updated

**Comment by rdmsr** [comment]
> how can i build and run from src because release is not updated

I'll try to add a Github actions CI to get releases

**Comment by circutrider21** [comment]
Have you thought of using a build system such as meson. Or using IO-APIC / LAPIC instead of the old PIC. I could help you if you are interested in both of those things!

**Reply by rdmsr** [reply]
I am using meson with ninja, and I'll see about the PIC!

**Reply by circutrider21** [reply]
Oh, I didn't know you had meson support already.

**Comment by circutrider21** [comment]
Where did you get the font in your kernel from?
I was just wondering out of curiosity.

**Reply by rdmsr** [reply]
I just took it from another repository :-) I'll probably add PSF support when I'm done fixing paging!

**Comment by Smart6502** [comment]
Hello, I'm Xenon6502, working on drivers for emeraldOS!!

**Reply by DamieFC** [reply]
Yay!

**Comment by DamieFC** [comment]
I actually have a question, is there an input function or something. Ok 2 questions, what is the function for the background color? 

Actually, is it the VBE_clear_screen() function? I think it is. THIS QUESTION DOES NOT NEED ATTENTION, but the first one does.

**Reply by DamieFC** [reply]
Also I like independent OS's but I think this is one of the coolest.

**Reply by DamieFC** [reply]
Not sure why

**Comment by rdmsr** [comment]
> I actually have a question, is there an input function or something. Ok 2 questions, what is the function for the background color? 
> 
> Actually, is it the VBE_clear_screen() function? I think it is. THIS QUESTION DOES NOT NEED ATTENTION, but the first one does.

Input function? What do you mean?

**Reply by DamieFC** [reply]
> Input function? What do you mean?

Like if you can ask, say, "what's your name?" and then allow the user to put in their name, and store it as a variable or something.

**Reply by rdmsr** [reply]
yeah, maybe I could do that. Just keep in mind that I don't want to make a shell at kernel level though.

**Reply by DamieFC** [reply]
@Abb1x I figured that, I thought it could be used in security or something, for the kernel.

**Comment by rdmsr** [comment]
Btw I'm going to start refactoring and reorganizing stuff so don't expect much functionality

**Comment by circutrider21** [comment]
Why is VMM Init commented out in main.c?

**Comment by rdmsr** [comment]
> Why is VMM Init commented out in main.c?

Because it crashes...

**Reply by circutrider21** [reply]
Yea, I noticed that when I un-commented it. I have done some digging and it seems like the problem is an Invalid-Opcode exception generated from the for loop.

**Reply by circutrider21** [reply]
Bochs gives me a different error. It generates a Double Fault and when I look at the logs, it says "canonical read violation"

**Reply by circutrider21** [reply]
Overall, I think the culprits are the for loops in VMM_init(). Think Clang is doing some weird optimization. Anyways, I'll look into it and see what the problem is. Also, have you tried using kvm in qemu, or compilng with GCC. Both of which have fixed my problems before.

*NOTE: Addresses on x86_64 systems only use the first 48 bits, while the bits 48 - 64 are 0. In the case of bochs, it checks to make sure that these bits are 0, otherwise it generates a "canonical read violation". I suspect that the invalid-opcode exception is qemu's way of saying the same thing*

**Reply by rdmsr** [reply]
The problem seems to be because `bitmap_size = 0` in my PMM, hence, the pmm_allocate_zero returns NULL and I am moving this NULL into cr3 = crash

**Reply by circutrider21** [reply]
Oh, I never thought to look in the pmm, but that does seem like a reason for a crash.
